### PR TITLE
[MM -55219] Create initial project structure

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Second time opening this, it's a bit of half way work but it should give us the chance to work on the project on several branches. First one was a bit of glitched because I accidentally pushed this branch and GitHub made it default branch so had to restore `master` and so on.
 
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55219

### CODE REVIEW

File: server/log.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	// these are to avoid any possible panic scenarios
+	if len(items) == 0 {
+		return nil
+	} else if len(items) == 1 {
+		m.api.LogInfo(fmt.Sprintf("%v", items[0]))
+		return nil
+	}
+
+	// we get the message and then remove the key value pair from the items slice
+	// then the removedd items will still be added to the list
```
Comment (agnivade) [inline server/log.go:39]: Can you add some sample log lines here? It's a bit hard to imagine how the log appears.

File: go.mod [hunk 2]
```diff
@@ -1,53 +1,101 @@
 module github.com/mattermost/mattermost-plugin-starter-template
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply go.mod:1]: Oh yea, we can change it. Totally missed that!

File: server/configuration.go [hunk 3]
```diff
@@ -31,16 +80,22 @@ func (c *configuration) Clone() *configuration {
...
+	defer p.configurationLock.Unlock()
 
 	if p.configuration == nil {
-		return &configuration{}
+		p.configuration = new(configuration)
+		p.configuration.SetDefaults()
 	}
 
-	return p.configuration
+	return p.configuration.Clone()
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply server/configuration.go:91]: Right, so we need to change clone implementation first as it doesn't make a deep copy. For this specific method, I think we still need to clone as the data will be all pointers.

File: server/configuration.go [hunk 4]
```diff
@@ -59,25 +114,89 @@ func (p *Plugin) setConfiguration(configuration *configuration) {
...
+
+	// Setting defaults prevents errors in case the plugin is updated after a new
+	// setting has been added. In this case the default value will be used.
+	cfg.SetDefaults()
+
+	if err := cfg.IsValid(); err != nil {
+		appErr := model.NewAppError("saveConfig", "app.save_config.error", nil, "", http.StatusBadRequest)
+		appErr.Message = err.Error()
+		appErr.SkipTranslation = true
+		return nil, appErr
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply server/configuration.go:188]: Sure 👍

File: server/plugin.go [hunk 5]
```diff
@@ -4,25 +4,84 @@ import (
...
+	// check if cluster is enabled
+	if p.isHA() {
+		// TODO(isacikgoz): get cluster info
+		p.API.LogWarn("cluster meterics is not enabled")
+	}
+
+	// initiate local tsdb
+	p.tsdbLock.Lock()
+	defer p.tsdbLock.Unlock()
+	p.db, err = tsdb.Open(*p.configuration.DBPath, logger, nil, &tsdb.Options{
```
Comment (agnivade) [inline server/plugin.go:66]: Does it create the dbpath if the directory doesn't exist? Or do we need to create it ourselves?

File: server/log_test.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+func TestLog(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ml := mocks.NewMockLogger(ctrl)
+	lg := &metricsLogger{api: ml}
+
+	t.Run("should not panic on a single line", func(t *testing.T) {
```
Comment (agnivade) [inline server/log_test.go:18]: Let's also add another case with `lg.Log()`

File: server/log.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	var msgIndex int
+	for i, item := range items {
+		if s, ok := item.(string); ok && s == "msg" {
+			msgIndex = i
+			break
+		}
+	}
+
+	// we get the message and then remove the key value pair from the items slice
+	// then the removedd items will still be added to the list
```
Comment (agnivade) [inline server/log.go:58]: Typo at removed

File: server/log_test.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+		ml.EXPECT().LogError("a log")
+		err := lg.Log("msg", "a log", "level", "error")
+		require.NoError(t, err)
+	})
+
+	t.Run("shuffle order on error log", func(t *testing.T) {
+		ml.EXPECT().LogError("a log")
+		err := lg.Log("level", "error", "msg", "a log")
+		require.NoError(t, err)
+	})
```
Comment (agnivade) [inline server/log_test.go:39]: Let's also add another case with a key other than `level` or `msg` and ensure that gets logged as well.

File: server/plugin.go (hunk 5 above)
Comment (isacikgoz) [reply server/plugin.go:66]: It creates automatically. Even if it's empty, it creates `workdir/data`. But it *may* mixed up with `mattermost/data`, hence I wanted to join with `PluginName`.

**Review by agnivade — APPROVED** [review]
Nicely done!

**Pull request merged**